management/src/main/java/org/ehcache/management/registry/DefaultCollectorService.java

```diff
@@ -0,0 +1,215 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultCollectorService.class);
+
+  private final AtomicBoolean collectingStats = new AtomicBoolean();
+  private volatile ScheduledFuture<?> task;
+
+  private final ConcurrentMap<String, StatisticQuery.Builder> selectedStatsPerCapability = new ConcurrentHashMap<String, StatisticQuery.Builder>();
+  private final MessageConsumer messageConsumer;
+
+  private volatile TimeSourceService timeSource;
+  private volatile ManagementRegistryService managementRegistry;
+  private volatile ScheduledExecutorService scheduledExecutorService;
+  private volatile CacheManagerProviderService cacheManagerProviderService;
+  private volatile ManagementRegistryServiceConfiguration configuration;
+
+  public DefaultCollectorService(MessageConsumer messageConsumer) {
+    this.messageConsumer = messageConsumer;
+  }
+
+  @Override
+  public void start(ServiceProvider serviceProvider) {
+    timeSource = serviceProvider.getService(TimeSourceService.class);
+    managementRegistry = serviceProvider.getService(ManagementRegistryService.class);
+    configuration = managementRegistry.getConfiguration();
+    cacheManagerProviderService = serviceProvider.getService(CacheManagerProviderService.class);
+    scheduledExecutorService = serviceProvider.getService(ExecutionService.class).getScheduledExecutor(DefaultCollectorService.class.getSimpleName());
+
+    cacheManagerProviderService.getCacheManager().registerListener(this);
+  }
+
+  @Override
+  public void stop() {
+    stopStatisticCollector();
+    shutdownNow(scheduledExecutorService);
+  }
+
+  @Override
+  public void cacheAdded(String alias, Cache<?, ?> cache) {
+    messageConsumer.accept(new DefaultMessage(
+        timeSource.getTimeSource().getTimeMillis(),
+        new ContextualNotification(
+            configuration.getContext().with("cacheName", alias),
+            EhcacheNotification.CACHE_ADDED.name())));
+  }
+
+  @Override
+  public void cacheRemoved(String alias, Cache<?, ?> cache) {
+    messageConsumer.accept(new DefaultMessage(
+        timeSource.getTimeSource().getTimeMillis(),
+        new ContextualNotification(
+            configuration.getContext().with("cacheName", alias),
+            EhcacheNotification.CACHE_REMOVED.name())));
+  }
+
+  @Override
+  @SuppressFBWarnings("SF_SWITCH_NO_DEFAULT")
+  public void stateTransition(Status from, Status to) {
+    switch (to) {
+
+      case AVAILABLE:
+        managementRegistry.register(this);
+
+        messageConsumer.accept(new DefaultMessage(
+            timeSource.getTimeSource().getTimeMillis(),
+            new ContextualNotification(
+                configuration.getContext(),
+                EhcacheNotification.CACHE_MANAGER_AVAILABLE.name())));
+        break;
+
+      case MAINTENANCE:
+        messageConsumer.accept(new DefaultMessage(
+            timeSource.getTimeSource().getTimeMillis(),
```

Comment:
Why are we always going back to the services to get stuff?  We can probably store the `TimeSource` and `CacheManager` directly, no?